### ISSUE TITLE
v0.12.0-beta1: Error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: tls: first record does not look like a TLS handshake

### ISSUE THREAD

**Issue by mcalhoun** [hashicorp/terraform#20714]
### Terraform Version
v0.12.0-beta1

### Debug Output
-----------------------------------------------------------------------------
2019/03/15 12:54:44 [DEBUG] checking for provider in "."
2019/03/15 12:54:44 [DEBUG] checking for provider in "/Users/matt/bin"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-azurerm_v1.17.0-4-g4ed5a4b3_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-google_v1.19.1-3-g59efc8b9_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-local_v1.1.0-4-g9b9be32_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-null_v1.0.0-4-g0b90acb_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-random_v2.0.0-4-gb282401_x4"
2019/03/15 12:54:44 [DEBUG] found provider "terraform-provider-template_v1.0.0-4-g17c9029_x4"
2019/03/15 12:54:44 [DEBUG] checking for provider in ".terraform/plugins/darwin_amd64"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "aws", "1.40.0-5-g4f24bc8b7", "/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "azurerm", "1.17.0-4-g4ed5a4b3", "/Users/matt/bin/terraform-provider-azurerm_v1.17.0-4-g4ed5a4b3_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "google", "1.19.1-3-g59efc8b9", "/Users/matt/bin/terraform-provider-google_v1.19.1-3-g59efc8b9_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "local", "1.1.0-4-g9b9be32", "/Users/matt/bin/terraform-provider-local_v1.1.0-4-g9b9be32_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "null", "1.0.0-4-g0b90acb", "/Users/matt/bin/terraform-provider-null_v1.0.0-4-g0b90acb_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "random", "2.0.0-4-gb282401", "/Users/matt/bin/terraform-provider-random_v2.0.0-4-gb282401_x4"
2019/03/15 12:54:44 [DEBUG] found valid plugin: "template", "1.0.0-4-g17c9029", "/Users/matt/bin/terraform-provider-template_v1.0.0-4-g17c9029_x4"
2019/03/15 12:54:44 [DEBUG] checking for provisioner in "."
2019/03/15 12:54:44 [DEBUG] checking for provisioner in "/Users/matt/bin"
2019/03/15 12:54:44 [DEBUG] checking for provisioner in ".terraform/plugins/darwin_amd64"
2019/03/15 12:54:44 [TRACE] configupgrade: Analyzing "main.tf"
2019/03/15 12:54:44 [TRACE] Provider block requires provider "aws"
2019/03/15 12:54:44 [TRACE] configupgrade: Analyzing "outputs.tf"
2019/03/15 12:54:44 [TRACE] configupgrade: Analyzing "vars.tf"
2019/03/15 12:54:44 [TRACE] Fetching schema from provider "aws"
2019-03-15T12:54:44.391-0700 [INFO]  plugin: configuring client automatic mTLS
2019-03-15T12:54:44.412-0700 [DEBUG] plugin: starting plugin: path=/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4 args=[/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4]
2019-03-15T12:54:44.415-0700 [DEBUG] plugin: plugin started: path=/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4 pid=31419
2019-03-15T12:54:44.415-0700 [DEBUG] plugin: waiting for RPC address: path=/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4
2019-03-15T12:54:44.427-0700 [DEBUG] plugin: using plugin: version=5
2019-03-15T12:54:44.427-0700 [DEBUG] plugin.terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4: plugin address: address=/var/folders/t5/8_hp6zgn6p70gk47w5xjkykw0000gn/T/plugin511833614 network=unix timestamp=2019-03-15T12:54:44.427-0700
2019/03/15 12:54:44 [TRACE] GRPCProvider: GetSchema

Error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: tls: first record does not look like a TLS handshake"

2019-03-15T12:54:46.428-0700 [WARN]  plugin: plugin failed to exit gracefully
2019-03-15T12:54:46.430-0700 [DEBUG] plugin: plugin process exited: path=/Users/matt/bin/terraform-provider-aws_v1.40.0-5-g4f24bc8b7_x4 pid=31419 error="signal: killed"

### Expected Behavior
Expected module to be upgraded to v0.12.0 syntax

### Actual Behavior
Error occurred

### Steps to Reproduce
1. `terraform init`
1. `terraform 0.12upgrade`

**Comment by apparentlymart** [comment]
Hi @mcalhoun! Thanks for reporting this.

From the error message there it looks like you might be using an `aws` provider plugin build from one of the alpha releases. Unfortunately some changes we made since the alphas changed some details of the plugin protocol, so the alpha builds of providers are not compatible with beta1 of Core.

If you remove that AWS provider binary and install [a dev snapshot](http://terraform-0.12.0-dev-snapshots.s3-website-us-west-2.amazonaws.com/terraform-provider-aws/) instead, I think this problem should go away.

Sorry for the weird error message here... one of the changes we made was to how the TLS settings are configured so it looks like the startup is failing in the TLS handshake code, before Terraform's own code is able to detect the incompatibility and return a better error message.
